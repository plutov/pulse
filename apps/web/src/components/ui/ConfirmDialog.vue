<template>
  <q-dialog :model-value="modelValue" persistent @update:model-value="emit('update:modelValue', $event)">
    <q-card style="min-width: 350px">
      <q-card-section>
        <div class="text-h6">{{ title }}</div>
      </q-card-section>

      <q-card-section class="q-pt-none">
        {{ message }}
      </q-card-section>

      <q-card-actions align="right" class="text-primary">
        <q-btn flat :label="cancelLabel" @click="onCancel" />
        <q-btn 
          flat 
          :label="confirmLabel" 
          :color="confirmColor"
          @click="onConfirm" 
          :loading="loading"
        />
      </q-card-actions>
    </q-card>
  </q-dialog>
</template>

<script setup lang="ts">
interface Props {
  modelValue: boolean;
  title: string;
  message: string;
  confirmLabel?: string;
  cancelLabel?: string;
  confirmColor?: string;
  loading?: boolean;
}

withDefaults(defineProps<Props>(), {
  confirmLabel: "Confirm",
  cancelLabel: "Cancel",
  confirmColor: "negative",
  loading: false,
});

const emit = defineEmits<{
  "update:modelValue": [value: boolean];
  confirm: [];
  cancel: [];
}>();

const onConfirm = () => {
  emit("confirm");
};

const onCancel = () => {
  emit("cancel");
  emit("update:modelValue", false);
};
</script>