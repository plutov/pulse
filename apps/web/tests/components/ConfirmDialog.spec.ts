import { mount } from "@vue/test-utils";
import { describe, it, expect } from "vitest";
import ConfirmDialog from "../../src/components/ui/ConfirmDialog.vue";

describe("ConfirmDialog.vue", () => {
  it("renders with default props", () => {
    const wrapper = mount(ConfirmDialog, {
      props: {
        modelValue: true,
        title: "Test Title",
        message: "Test Message",
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("renders with custom props", () => {
    const wrapper = mount(ConfirmDialog, {
      props: {
        modelValue: true,
        title: "Custom Title",
        message: "Custom Message",
        confirmLabel: "Yes",
        cancelLabel: "No",
        confirmColor: "positive",
        loading: true,
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it("emits confirm event when confirm button is clicked", async () => {
    const wrapper = mount(ConfirmDialog, {
      props: {
        modelValue: true,
        title: "Test",
        message: "Test",
      },
    });

    await wrapper.findAll("q-btn")[1]?.trigger("click");

    expect(wrapper.emitted()).toHaveProperty("confirm");
  });

  it("emits cancel and update:modelValue events when cancel button is clicked", async () => {
    const wrapper = mount(ConfirmDialog, {
      props: {
        modelValue: true,
        title: "Test",
        message: "Test",
      },
    });

    await wrapper.findAll("q-btn")[0]?.trigger("click");

    expect(wrapper.emitted()).toHaveProperty("cancel");
    expect(wrapper.emitted()).toHaveProperty("update:modelValue");
    expect(wrapper.emitted("update:modelValue")![0]).toEqual([false]);
  });

  it("shows loading state on confirm button", () => {
    const wrapper = mount(ConfirmDialog, {
      props: {
        modelValue: true,
        title: "Test",
        message: "Test",
        loading: true,
      },
    });

    const confirmButton = wrapper.findAll("q-btn")[1];
    expect(confirmButton?.attributes("loading")).toBeDefined();
  });
});
