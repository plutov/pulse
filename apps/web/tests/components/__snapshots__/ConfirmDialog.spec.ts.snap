// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConfirmDialog.vue > renders with custom props 1`] = `
"<q-dialog model-value="true" persistent="">
  <q-card style="min-width: 350px;">
    <q-card-section>
      <div class="text-h6">Custom Title</div>
    </q-card-section>
    <q-card-section class="q-pt-none">Custom Message</q-card-section>
    <q-card-actions align="right" class="text-primary">
      <q-btn flat="" label="No"></q-btn>
      <q-btn flat="" label="Yes" color="positive" loading="true"></q-btn>
    </q-card-actions>
  </q-card>
</q-dialog>"
`;

exports[`ConfirmDialog.vue > renders with default props 1`] = `
"<q-dialog model-value="true" persistent="">
  <q-card style="min-width: 350px;">
    <q-card-section>
      <div class="text-h6">Test Title</div>
    </q-card-section>
    <q-card-section class="q-pt-none">Test Message</q-card-section>
    <q-card-actions align="right" class="text-primary">
      <q-btn flat="" label="Cancel"></q-btn>
      <q-btn flat="" label="Confirm" color="negative" loading="false"></q-btn>
    </q-card-actions>
  </q-card>
</q-dialog>"
`;
